### PR TITLE
Add ChromeEventTracer

### PR DESCRIPTION
**Pull request by grfrost**
These are the basic files allowing us to create Chrome compatible json tracing files. 

To enable tracing 
`-Dtornado.chrome.event.tracer.enabled=True`

The default file is "chrome.json". To select an alternate file 
`-Dtornado.chrome.event.tracer.filename=cascade-events.json`

[See this doc](https://docs.google.com/document/d/1Q9HtGveV78_Ilw_wpO-5eqSAeKh3MUeG8UYhO4RT-QQ/edit?usp=sharing)

### CODE REVIEW

**Comment by jjfumero** [comment]
Thanks, @grfrost for this feature. 

I have a few questions:

1. It seems that the option is enabled by default. If `-Dtornado.chrome.event.tracer.enabled=False` is still generating a JSON file with the following content:
```json
{
 "traceEvents":[
  {
   "args":{
    "name":"Tornado"
   }
   ,
   "ph":"M",
   "pid":10642,
   "tid":1,
   "name":"tornadovm",
   "sort_index":1
  }
  ]
 }
```

2. When running with this option on:

```bash
$ tornado -Dtornado.chrome.event.tracer.enabled=True uk.ac.manchester.tornado.examples.compute.NBody 256
```
I can't see the copy-IN 

```json
{
 "traceEvents":[
  {
   "args":{
    "name":"Tornado"
   }
   ,
   "ph":"M",
   "pid":10801,
   "tid":1,
   "name":"tornadovm",
   "sort_index":1
  }
  ,{"ph":"X","name":"nbody sequential","cat":"exec","pid":10801,"tid":1,"ts":-11508,"dur":8764
  }
  ,{"ph":"X","name":"nbody sequential","cat":"exec","pid":10801,"tid":1,"ts":58109,"dur":1366
  }
  ,{"ph":"X","name":"nbody sequential","cat":"exec","pid":10801,"tid":1,"ts":67255,"dur":5423
  }
  ,{"ph":"X","name":"nbody sequential","cat":"exec","pid":10801,"tid":1,"ts":77106,"dur":1238
  }
  ,{"ph":"X","name":"nbody sequential","cat":"exec","pid":10801,"tid":1,"ts":83572,"dur":1154
  }
  ,{"ph":"X","name":"nbody sequential","cat":"exec","pid":10801,"tid":1,"ts":88852,"dur":1623
  }
...
```

Is this process manually? Meaning that I should change the user application (in this case NBODY) to pass the start-stop from the copies to/from the device? 

Something similar to this?

```java
ChromeEventTracer.enqueueTaskIfEnabled("nbody copy H->D", start, end);
```

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
+package uk.ac.manchester.tornado.api.profiler;
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:1]: Add license. This should be the same as any other file from the API module. 
Example: 
https://github.com/beehive-lab/TornadoVM/blob/master/tornado-api/src/main/java/uk/ac/manchester/tornado/api/GridTask.java#L1-L41

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            VMManagement mgmt = (VMManagement) jvm.get(runtime);
+            java.lang.reflect.Method pid_method = mgmt.getClass().getDeclaredMethod("getProcessId");
+            pid_method.setAccessible(true);
+            pid = (Integer) pid_method.invoke(mgmt);
+        } catch (NoSuchFieldException | IllegalAccessException | NoSuchMethodException | InvocationTargetException nsf) {
+            // well we tried. Lets just use the default 0
+        }
+        objectStart();
+        arrayStart("traceEvents");
+        object(() -> {
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:32]: This code answers my first question. So the fact that the API calls this interface automatically sets the JSON file.

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+public class ChromeEventJSonWriter extends JSonWriter<ChromeEventJSonWriter> {
+    static private int pid = 0;
+    ContentWriter NO_ARGS = null;
+
+    ChromeEventJSonWriter() {
+        super();
+        // Java 9 we could use long pid = ProcessHandle.current().pid();
+        // For java 8 we must go through some hoops.
+        try {
+            RuntimeMXBean runtime = ManagementFactory.getRuntimeMXBean();
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:20]: When compiling with Corretto-11 I get the following error:

```bash
Compilation failure: 
[ERROR] /tmp/tornadovm-installer/TornadoVM-Amazon-Corretto11/TornadoVM/tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:[3,10] error: package sun.management is not visible
[ERROR]   (package sun.management is declared in module java.management, but module tornado.api does not read it)
[ERROR] /tmp/tornadovm-installer/TornadoVM-Amazon-Corretto11/TornadoVM/tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:[6,16] error: package java.lang.management is not visible
[ERROR]   (package java.lang.management is declared in module java.management, but module tornado.api does not read it)
[ERROR] /tmp/tornadovm-installer/TornadoVM-Amazon-Corretto11/TornadoVM/tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:[7,16] error: package java.lang.management is not visible
[ERROR]   (package java.lang.management is declared in module java.management, but module tornado.api does not read it)
``` 

Is this code also valid using Java >= 8

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
+package uk.ac.manchester.tornado.api.profiler;
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java:1]: Apply license (TornadoVM-API license)

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+import java.io.*;
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java:3]: Unfold all imports according to the formatter

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java [hunk 6]
```diff
@@ -0,0 +1,258 @@
+package uk.ac.manchester.tornado.api.profiler;
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java:1]: Apply license

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+import java.io.*;
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java:3]: Unfold all imports

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Map;
+
+public class ChromeEventTracer {
```
Comment (jjfumero) [inline tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java:6]: Is possible this class implements the profiler interface?

**Review by jjfumero — CHANGES_REQUESTED** [review]
This is looking good. 
I added a few questions/comments. Please contact me for questions. 

Thank you!

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java (hunk 1 above)
Comment (grfrost) [reply tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:1]: Will do

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java (hunk 3 above)
Comment (grfrost) [reply tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventJSonWriter.java:20]: I will pull this bean code.  Will just use 0 for pid.

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java (hunk 6 above)
Comment (grfrost) [reply tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java:1]: Hmm code formatter should do this. But yes.  Agree 100%

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java (hunk 7 above)
Comment (grfrost) [reply tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/JSonWriter.java:3]: Yep

File: tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java (hunk 8 above)
Comment (grfrost) [reply tornado-api/src/main/java/uk/ac/manchester/tornado/api/profiler/ChromeEventTracer.java:6]: No.  Started with that, but many of the API's don't make sense.

**Comment by jjfumero** [comment]
Looking good. @grfrost , can you add an example in the `examples` module with all data copies plus kernel time? This is to show how to use this API. 

This example should be in the `uk.ac.manchester.tornado.examples.profiler` package (create one).

**Review by jjfumero — APPROVED** [review]
Tested with JDK8 and Corretto 11.

**Comment by grfrost** [comment]
Re : - "can you add an example in the examples module with all data copies
plus kernel time? This is to show how to use this API."

No. I Can't.  This check in (as I have mentioned a few times now......), is
limited to the actual Tracer and associated Json writer. I deliberately
made no changes to any other TorndadoVM classe, because event handling is
completely broken.

I can't collect kernel time or data move time until we merge this
capability into the rest of the code ( OCLDeviceContext and/or the
OCLCommanQueue and/or the OCLEventsWrapper) depending on the solution we
decide to fix broken event handling.

If you want to wait until I can check in an example, let me know.  I will
withdraw the pull request and represent when event handling is fixed.

But as we mentioned yesterday event handling fix might be a BIG change.

**Comment by jjfumero** [comment]
True. I got confused with the way NBody collects the start-stop. In the case of copies, we will need this directly from the OpenCL events. Let's merge this and prepare the new design for fixing the events.

**Review by kotselidis — APPROVED** [review]

**Pull request merged**